File: pocs/code-review-demo/agents/security_agent.py
from beeai_framework.agents.experimental import RequirementAgent
from beeai_framework.agents.experimental.requirements.conditional import ConditionalRequirement
from beeai_framework.memory import UnconstrainedMemory
from beeai_framework.tools import Tool
from beeai_framework.tools.think import ThinkTool
from core.backend import get_llm


def create_security_agent() -> RequirementAgent:
    return RequirementAgent(
        llm=get_llm(),
        memory=UnconstrainedMemory(),
        tools=[ThinkTool()],
        requirements=[
            ConditionalRequirement(ThinkTool, force_at_step=1),
            ConditionalRequirement(ThinkTool, force_after=Tool, consecutive_allowed=False),
        ],
        role="security_reviewer",
        instructions=(
            "You are a security engineer performing a thorough code security audit. "
            "Focus on: SQL injection, command injection, exposed secrets, eval() on external data, "
            "and unsanitized inputs. Always think before and after each finding. "
            "Mark every issue as critical, warning, or info."
        ),
    )

File: pocs/code-review-demo-ts/src/agents/securityAgent.ts
import { RequirementAgent } from "beeai-framework/agents/requirement/agent";
import { ConditionalRequirement } from "beeai-framework/agents/requirement/requirements/conditional";
import { Tool } from "beeai-framework/tools/base";
import { ThinkTool } from "beeai-framework/tools/think";

import { getLLM } from "../core/backend.js";

export async function createSecurityAgent() {
  return new RequirementAgent({
    llm: await getLLM(),
    tools: [new ThinkTool()],
    instructions:
      `You are a security expert. Review the submitted code for SQL injection, eval() misuse, exposed secrets,
      command injection, and unsanitized inputs. Be thorough and conservative — flag anything suspicious. 
      Return a structured list of findings with severity: CRITICAL, WARNING, or INFO.`,
    requirements: [
      new ConditionalRequirement(ThinkTool, { forceAtStep: 1, maxInvocations: 5 }),
      new ConditionalRequirement(ThinkTool, { forceAfter: [Tool], consecutiveAllowed: false }),
    ],
  });
}
